llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,935 @@
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/LineIterator.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Transforms/IPO.h"
+#include "llvm/Transforms/IPO/GlobalDCE.h"
+#include "llvm/Transforms/IPO/Internalize.h"
+#include "llvm/Transforms/IPO/StripDeadPrototypes.h"
+#include "llvm/Transforms/IPO/StripSymbols.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include "llvm/Transforms/Utils/SYCLUtils.h"
+
+#include <algorithm>
+#include <map>
+#include <utility>
+#include <variant>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "sycl_module_split"
+
+namespace {
+// Identifying name for global scope
+constexpr char GLOBAL_SCOPE_NAME[] = "<GLOBAL>";
+constexpr char SYCL_SCOPE_NAME[] = "<SYCL>";
+
+EntryPointsGroupScope selectDeviceCodeGroupScope(const Module &M,
+                                                 IRSplitMode Mode,
+                                                 bool AutoSplitIsGlobalScope) {
+  switch (Mode) {
+  case SPLIT_PER_TU:
+    return Scope_PerModule;
+
+  case SPLIT_PER_KERNEL:
+    return Scope_PerKernel;
+
+  case SPLIT_AUTO: {
+    if (AutoSplitIsGlobalScope)
+      return Scope_Global;
+
+    // At the moment, we assume that per-source split is the best way of
+    // splitting device code and can always be used except for cases handled
+    // above.
+    return Scope_PerModule;
+  }
+
+  case SPLIT_NONE:
+    return Scope_Global;
+  }
+
+  llvm_unreachable("unsupported split mode");
+}
+
+// Return true if the function is a SPIRV or SYCL builtin, e.g.
+// _Z28__spirv_GlobalInvocationId_xv
+bool isSpirvSyclBuiltin(StringRef FName) {
+  if (!FName.consume_front("_Z"))
+    return false;
+  // now skip the digits
+  FName = FName.drop_while([](char C) { return std::isdigit(C); });
+
+  return FName.starts_with("__spirv_") || FName.starts_with("__sycl_");
+}
+
+// Return true if the function name starts with "__builtin_"
+bool isGenericBuiltin(StringRef FName) {
+  return FName.starts_with("__builtin_");
+}
+
+bool isKernel(const Function &F) {
+  return F.getCallingConv() == CallingConv::SPIR_KERNEL ||
+         F.getCallingConv() == CallingConv::AMDGPU_KERNEL;
+}
+
+bool isEntryPoint(const Function &F, bool EmitOnlyKernelsAsEntryPoints) {
+  // Skip declarations, if any: they should not be included into a vector of
+  // entry points groups or otherwise we will end up with incorrectly generated
+  // list of symbols.
+  if (F.isDeclaration())
+    return false;
+
+  // Kernels are always considered to be entry points
+  if (isKernel(F))
+    return true;
+
+  if (!EmitOnlyKernelsAsEntryPoints) {
+    // If not disabled, SYCL_EXTERNAL functions with sycl-module-id attribute
+    // are also considered as entry points (except __spirv_* and __sycl_*
+    // functions)
+    return llvm::isSYCLExternalFunction(&F) &&
+           !isSpirvSyclBuiltin(F.getName()) && !isGenericBuiltin(F.getName());
+  }
+
+  // Even if we are emitting only kernels as entry points, virtual functions
```

Comment:
Comments like this reference various other features and I'm concerned that the upstream folks will either ask for their implementation, or at least their documentation. Not sure what is the right approach here (leave as-is, document better, don't upstream for now, etc.)

Reply:
I would prefer upstreaming minimum set of features. Once the code is available upstream, we should be able to add new features incrementally.

Reply:
Removed that.